test/data_type/test_custom_type.cpp

```diff
@@ -0,0 +1,464 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2024, Advanced Micro Devices, Inc. All rights reserved.
+
+#include "gtest/gtest.h"
+#include "ck/utility/data_type.hpp"
+#include "ck/utility/type_convert.hpp"
+
+using ck::bf8_t;
+using ck::bhalf_t;
+using ck::f8_t;
+using ck::half_t;
+using ck::Number;
+using ck::type_convert;
+using ck::vector_type;
+
+TEST(Custom_bool, TestSize)
+{
+    struct custom_bool_t
+    {
+        bool data;
+    };
+    ASSERT_EQ(sizeof(bool), 1);
+    ASSERT_EQ(sizeof(custom_bool_t), 1);
+    ASSERT_EQ(sizeof(vector_type<bool, 2>), 2);
+    ASSERT_EQ(sizeof(vector_type<custom_bool_t, 2>), 2);
+    ASSERT_EQ(sizeof(vector_type<bool, 4>), 4);
+    ASSERT_EQ(sizeof(vector_type<custom_bool_t, 4>), 4);
+    ASSERT_EQ(sizeof(vector_type<bool, 8>), 8);
+    ASSERT_EQ(sizeof(vector_type<custom_bool_t, 8>), 8);
+    ASSERT_EQ(sizeof(vector_type<bool, 16>), 16);
+    ASSERT_EQ(sizeof(vector_type<custom_bool_t, 16>), 16);
+    ASSERT_EQ(sizeof(vector_type<bool, 32>), 32);
+    ASSERT_EQ(sizeof(vector_type<custom_bool_t, 32>), 32);
+    ASSERT_EQ(sizeof(vector_type<bool, 64>), 64);
+    ASSERT_EQ(sizeof(vector_type<custom_bool_t, 64>), 64);
+}
+
+TEST(Custom_bool, TestAsType)
+{
+    struct custom_bool_t
+    {
+        using type = bool;
+        type data;
+        custom_bool_t() : data{type{}} {}
+        custom_bool_t(type init) : data{init} {}
+    };
+
+    // test size
+    const int size             = 4;
+    std::vector<bool> test_vec = {false, true, false, true};
+    // reference vector
+    vector_type<custom_bool_t, size> right_vec;
```

Comment:
You could also check here if all values were properly default initialized. They all should have value equal to the result of calling default constructor `custom_bool_t() : data{type{}} {}` . Similar for all other test cases.

Reply:
If I understand your idea correctly, that's exactly what we check in these tests: we create a std vector with some values (test_vec); assign values to a vector type (right_vec) element-wise; create another vector (left_vec) using the default constructor; compare the new vector (left_vec) with the reference vector (test_vec). This way we can check both default constructor and AsType method. Please let me know if I missed something!

Reply:
I meant that after this line `vector_type<custom_bool_t, size> right_vec;` all values of `right_vec` should have been _default initialized_ - that is have a default value after calling default constructor - (usually zero for numeric types, boolean is false). Since we have vector of values, then I'd expect that all elements after this line have default value.

Reply:
Added in [46a3729](https://github.com/ROCm/composable_kernel/pull/1333/commits/46a3729fe996508a3c9c62f1b39b8ce4d428d4a3)